src/Microsoft.AspNet.OData.Shared/Query/ODataQueryOptions.cs

```diff
@@ -416,12 +426,20 @@ public virtual IQueryable ApplyTo(IQueryable query, ODataQuerySettings querySett
             if (pageSize > 0)
             {
                 bool resultsLimited;
-                result = LimitResults(result, pageSize, out resultsLimited);
+                result = LimitResults(result, pageSize, out resultsLimited);               
                 if (resultsLimited && InternalRequest.RequestUri != null && InternalRequest.RequestUri.IsAbsoluteUri &&
                     InternalRequest.Context.NextLink == null)
                 {
-                    Uri nextPageLink = InternalRequest.GetNextPageLink(pageSize);
-                    InternalRequest.Context.NextLink = nextPageLink;
+                    DefaultQuerySettings settings = Context.RequestContainer.GetRequiredService<DefaultQuerySettings>();
+                    if (settings.EnableSkipToken)
+                    {
+                        string skipTokenValue = SkipTokenQueryOption.GetSkipTokenValue(result, this.Context.Model, OrderBy);
+                        InternalRequest.GetNextPageLink(pageSize, skipTokenValue);
+                    }
```

Comment:
We should move this to serialization instead of trying to get the last element here.